File: src/MatchingAlgorithm.js
import haversine from 'haversine-distance';
import { getPets, getCoords, getOwners } from '../service';

function typeOfAnimal(pet, owner) {
  if (pet.type === owner.type) {
    pet_list.push(pet);
  }
}

function smallChildren(pet, owner) {
  if (pet.smallChildren === 0 && owner.smallChildren === 1) {
    const index = pet_list.indexOf(pet);
    if (index !== -1) {
      pet_list.splice(index, 1);
    }
  }
}

function otherAnimals(pet, owner) {
  if (pet.otherAnimals === 0 && owner.otherPets === 1) {
    const index = pet_list.indexOf(pet);
    if (index !== -1) {
      pet_list.splice(index, 1);
    }
  }
}

function distanceLen(owner_ll, pet_ll) {
  var distances=haversine(owner_ll, pet_ll)
  //console.log("DISTANCE~~", distances)
  return distances
}

function maxDistance(lat_long_owner,lat_long_pet,owner, score) {
  const distance = distanceLen(lat_long_owner, lat_long_pet);

  if (distance < owner.travel) {
    score += 3;
  } else if (distance < owner.travel + 35) {
    score += 2;
  }

  return score;
    
}

function priceRange(pet, score, owner) {
    if (pet.price < owner.price) {
        score = score + 3;
    } else if (pet.price < owner.budget + 100) {
        score = score + 2;
    }
    return score;
}
    
function activityLevel(pet, score, owner) {
    if (owner.actLevel == pet.actLevel) {
        score = score + 3;
    }
    if (owner.actLevel + 1 == pet.actLevel) {
        score = score + 2;
    }
    if (owner.actLevel - 1 == pet.actLevel) {
        score = score + 2;
    }
    if (owner.actLevel + 2 == pet.actLevel) {
        score = score + 1;
    }
    if (owner.actLevel - 2 == pet.actLevel) {
        score = score + 1;
    }
    return score;
}

function age(pet, score, owner) {
    let pet_age;
    let owner_age;

    if (pet.age < 3) {
        pet_age = 0;
    } else if (pet.age > 7) {
        pet_age = 2;
    } else {
        pet_age = 1;
    }

    if (owner.age < 3) {
        owner_age = 0;
    } else if (owner.age > 7) {
        owner_age = 2;
    } else {
        owner_age = 1;
    }

    if (pet_age == owner_age) {
        score = score + 3;
    }
    if (pet_age + 1 == owner_age) {
        score = score + 1;
    }
    if (pet_age - 1 == owner_age) {
        score = score + 1;
    }
    return score;
}

// global pet list arrays
let pet_list = [];
let pet_high = [];
let pet_med = [];
let pet_low = [];

export async function match() {

  const owners = await getOwners();
  const owner = owners[0];

  const lat_long_owner = await getCoords(owner);

  // reset global arrays
  pet_list = []
  pet_high = []
  pet_med = [];
  pet_low = [];

  const pets = await getPets();

  for (let i = 0; i < pets.length; i++) {
    // get next pet and initialize score
    let pet = pets[i];
    let score = 0;

    // check for type of animal
    typeOfAnimal(pet, owner);
    if (pet_list.includes(pet, owner)) {
      // small children and other animals check
      if (pet_list.includes(pet)) {
        smallChildren(pet, owner);
        if (pet_list.includes(pet)) {
          otherAnimals(pet, owner);
        }
      }
    }

    // acccumulate score
    if (pet_list.includes(pet)) {
      score = age(pet, score, owner);
      score = activityLevel(pet, score, owner);
      score = priceRange(pet, score, owner);
      
      let lat_long_pet = await getCoords(pet);
      let final_score = maxDistance(lat_long_owner, lat_long_pet, owner, score);
      //console.log(pet.name, final_score);

      if (final_score < 5) {
        pet_low.push(pet);
      } else if (final_score > 9) {
        pet_high.push(pet);
      } else {
        pet_med.push(pet);
      }
    }
  }
    console.log("high:", pet_high);
    console.log("med:", pet_med);
    console.log("low:", pet_low);

    // const data = [pet_high, pet_med, pet_low];
    // console.log(data);
    return [pet_high, pet_med, pet_low];
}

File: Matching_algorithm.py
import json
from geopy.geocoders import Nominatim
import haversine as hs
from haversine import Unit


#JSON
x='{"Owner_Name":"Kali","travel":1000,"Location":"1866 East Shore Drive Maplewood Minnesota", "type":"Dog","otherPets":"Yes","smallChildren":"Yes","Size":"Large","Age":"Puppy","actLevel":1,"budget":500}'
y='{"Pet_Name":"Bronco","Location":"1325 31st street des moines Iowa", "type":"Dog","pet_breed":"Pit Bull","smallChildren":"Yes","otherAnimals":"Yes","Weight":"95","Age":9,"actLevel":2,"price":450}'
z='{"Pet_Name":"Perfect","Location":"1870 East shore dr maplewood mn 55109", "type":"Dog","pet_breed":"Pit Bull","smallChildren":"Yes","otherAnimals":"Yes","Weight":"95","Age":1,"actLevel":1,"price":450}'
a='{"Pet_Name":"BAD","Location":"352 gulfport court las vegas nv", "type":"Dog","pet_breed":"Pit Bull","smallChildren":"Yes","otherAnimals":"Yes","Weight":"10","Age":10,"actLevel":5,"price":1000}'
b='{"Pet_Name":"Tester","Location":"352 gulfport court las vegas nv", "type":"Dog","pet_breed":"Pit Bull","smallChildren":"Yes","otherAnimals":"Yes","Weight":"10","Age":10,"actLevel":5,"price":100000}'

owner = json.loads(x)
pet1 = json.loads(y)
pet2 = json.loads(a)
pet3 = json.loads(z)
pet4 = json.loads(b)

pets=[pet1,pet2,pet3,pet4]


pet_list=[]


def loc_distance(pet):
    geolocator=Nominatim(user_agent="MyApp")

    location_owner=geolocator.geocode(owner["Location"])
    lat=(location_owner.latitude)
    long=(location_owner.longitude)
    loc1=(lat,long)

    location_pet=geolocator.geocode(pet["Location"])
    latp=(location_pet.latitude)
    longp=(location_pet.longitude)
    loc2=(latp,longp)

    distance=hs.haversine(loc1,loc2, unit=Unit.MILES)
    return distance

def type_of_animal(pet):
    if pet["type"]==owner["type"]:
        pet_list.append(pet)

def small_children(pet):
    if pet["smallChildren"]=='No' and owner["smallChildren"]=="Yes":
        pet_list.remove(pet)

def other_animals(pet):
    if pet["otherAnimals"]=="No" and owner["otherPets"]=="Yes":
        pet_list.remove(pet)

def max_distacne(pet, score):
    distance=loc_distance(pet)
    if distance < owner["travel"]:
        score=score+3
    elif distance < owner["travel"]+35:
        score=score+2
    return score


def price_range(pet, score):
    if pet["price"]<owner["budget"]:
        score=score+3
    elif pet["price"]<owner["budget"]+100:
        score=score+2
    return score

        

def activity_level(pet, score):
    if owner["actLevel"]==pet["actLevel"]:
        score=score+3
    if owner["actLevel"]+1==pet["actLevel"]:
        score=score+2
    if owner["actLevel"]-1==pet["actLevel"]: 
        score=score+2
    if owner["actLevel"]+2==pet["actLevel"]:
        score=score+1
    if owner["actLevel"]-2==pet["actLevel"]: 
        score=score+1
    return score

def age(pet, score):
    if pet["Age"]<3:
        pet_age=0
    elif pet["Age"]>7:
        pet_age=2
    else:
        pet_age=1
    if owner["Age"]=="Puppy":
        owner_age=0
    elif owner["Age"]=="Senior":
        owner_age=2
    else:
        owner_age=1
    if pet_age==owner_age:
        score=score+3
    if pet_age+1==owner_age:
        score=score+1
    if pet_age-1==owner_age:
        score=score+1
    return score 


def main():
    pet_high=[]
    pet_med=[]
    pet_low=[]
    for pet in pets:
        score=0
        type_of_animal(pet)
        if pet in pet_list:
            small_children(pet)
        if pet in pet_list:
            other_animals(pet)
        if pet in pet_list:
            a=age(pet, score)
            b=activity_level(pet, a)
            c=price_range(pet, b)
            final_score=max_distacne(pet, c)
            print(pet["Pet_Name"],final_score)
            if final_score <5:
                pet_low.append(pet)
            elif final_score >9:
                pet_high.append(pet)
            else:
                pet_med.append(pet)
    print("high:",pet_high,"med:",pet_med,"low:",pet_low)
    
    
main()

#Notes:
#I need to figure out how I am looping through all the pets and calling the data. 


#Need an error code for if an non existant address is entered
